src/typed-ast/expr.h

```diff
@@ -0,0 +1,637 @@
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("if")
+        .field(condition)
+        .field(then_branch)
+        .optional_field(else_branch)
+        .finish();
+    }
+  };
+
+  /// This represents a while loop expression, e.g. `while(C) { E }`.
+  /// The expression evaluates to Unit.
+  class WhileExpr final : public Expr
+  {
+    ExprPtr condition;
+    ExprPtr body;
+
+  public:
+    explicit WhileExpr(SourceLocation location, ExprPtr condition, ExprPtr body)
+    : Expr(Kind::While, location),
+      condition(std::move(condition)),
+      body(std::move(body))
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::While;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("while").field(condition).field(body).finish();
+    }
+  };
+
+  /// This represents a `continue` statement.
+  class ContinueExpr final : public Expr
+  {
+  public:
+    explicit ContinueExpr(SourceLocation location)
+    : Expr(Kind::Continue, location)
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::Continue;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("continue").finish();
+    }
+  };
+
+  /// This represents a `break` statement.
+  class BreakExpr final : public Expr
+  {
+  public:
+    explicit BreakExpr(SourceLocation location) : Expr(Kind::Break, location) {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::Break;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("break").finish();
+    }
+  };
+
+  /// This represents a `return` statement.
+  class ReturnExpr final : public Expr
+  {
+    ExprPtr value;
+
+  public:
+    explicit ReturnExpr(SourceLocation location, ExprPtr value)
+    : Expr(Kind::Return, location), value(std::move(value))
+    {}
+    explicit ReturnExpr(SourceLocation location) : ReturnExpr(location, nullptr)
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::Return;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("return").optional_field(value).finish();
+    }
+  };
+
+  /// This represents a `yield` statement.
+  class YieldExpr final : public Expr
+  {
+    ExprPtr value;
+
+  public:
+    explicit YieldExpr(SourceLocation location, ExprPtr value)
+    : Expr(Kind::Yield, location), value(std::move(value))
+    {}
+
+    explicit YieldExpr(SourceLocation location) : YieldExpr(location, nullptr)
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::Yield;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("yield").field(value).finish();
+    }
+  };
+
+  /// This represents a call to an object, e.g. `E1(E2, E3)`.
+  ///
+  /// In the case of method calls, eg. `E.m()`, the receiver of the call is
+  /// `E.m`, a partial application of the method.
```

Comment:
Why do we have this to support method calls when we already have the `MethodCallExpr` node?

Reply:
If we have `x.f(y)`, we don't know yet if this is a field lookup, followed by a call the resulting object, or if it is a call to the method `f`. `MethodCallExpr` is used for operator calls, which aren't allowed to be fields.

 I'm not happy with it either, but that's what the untyped AST looks like right now.

Reply:
Ah, right, I have to handle that in the `parseCall` too and they're virtually identical. Perhaps we could common them up here even if the old ast doesn't, if they're semantically equivalent for all passes after this one.

Reply:
They aren't equivalent though. If `f` is a field, then the semantics of InvokeExpr and MethodCallExpr are different. Except we won't know what it is until late in the pipeline.

Reply:
Hm, ok, so we'll have to update the generator, because I'm treating both the same. There weren't any example of calling a field (or a method FWIW), so I was a bit lost. It should be simple to do that distinction, though, either as a new dialect operation or a variation on the existing one. We'll worry about that when we have the new ast.